iceoryx_posh/include/iceoryx_posh/internal/popo/trigger_state.inl

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2020 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IOX_POSH_POPO_TRIGGER_STATE_INL
+#define IOX_POSH_POPO_TRIGGER_STATE_INL
+
+namespace iox
+{
+namespace popo
+{
+template <typename T>
+inline void myCallback(void* const origin, TriggerState::Callback<void> callbackPtr) noexcept
+{
+    (*reinterpret_cast<TriggerState::Callback<T>>(callbackPtr))(reinterpret_cast<T*>(origin));
+}
+
+template <typename T>
+inline TriggerState::TriggerState(T* const origin, const uint64_t triggerId, const Callback<T> callback) noexcept
+    : m_origin(origin)
+    , m_originTypeHash(typeid(T).hash_code())
+    , m_triggerId(triggerId)
+    , m_callbackPtr(reinterpret_cast<Callback<void>>(callback))
+    , m_callback(myCallback<T>)
+{
+}
+
+template <typename T>
+inline bool TriggerState::doesOriginateFrom(T* const origin) const noexcept
+{
+    if (m_origin == nullptr)
+    {
+        return false;
+    }
+    return m_origin == origin;
+}
+
+template <typename T>
+inline T* TriggerState::getOrigin() noexcept
+{
+    if (m_originTypeHash != typeid(T).hash_code())
```

Comment:
If RTTI is not allowed why do we not compile the whole code base with `-fno-rtti`? Where is this information coming from? @budrus what is your opinion?

I can remove this but the only effect this would have is that the code and our public API is less safe to use!

Reply:
In general we should be careful with every new dependency and clarify if this is something you can get a certified implementation. I don't have a clear opinion on typeinfo. But it is a new dependency, or?

Reply:
Well, IIRC we had some corridor discuss about this long time ago. I wasn't aware that we're not using `-fno-rtti`. In the embedded world RTTI is in general discouraged to use due to the restricted resources. However, we should [check](https://stackoverflow.com/questions/579887/how-expensive-is-rtti). I know some places in the code bases where `dynamic_cast`'s could be handy.

Reply:
@budrus `typeinfo` is a new dependency! 
For the moment I would leave the code as it is since it improves the safety. If we would like to get rid of this we can do it easily since we just use `typeid` and `typeinfo` to verify at runtime that the user is requesting the same type they has provided when constructing the trigger.